src/libstd/ffi/c_str.rs

```diff
@@ -522,6 +522,37 @@ impl ToOwned for CStr {
     }
 }
 
+#[unstable(feature = "cstring_asref", reason = "recently added", issue = "0")]
+impl ops::Index<ops::RangeFull> for CString {
```

Comment:
OsString does the same thing, though that's intentionally an opaque type so the situation is slightly different in that slicing it is actually "impossible"...

I'm iffy on whether we should allow slicing into a CStr, just because it doesn't seem like a very common operation. It's also not necessarily clear whether you actually want to slice `as_bytes()` or `as_bytes_with_nul()`, and `&cstr[..1]` is impossible to implement without panicking if it doesn't cover the entire string.

This impl was meant as a conversion trait (the good old `&deref/&index[..]/AsRef/.as_slice()` bunch), the main advantage being that it isn't type-ambiguous. The discussion at #27729 might be a bit relevant there?